### DISCUSSION TITLE
EGSnrc on Windows Subsystem for Linux (WSL)

### DISCUSSION THREAD

**Discussion by crcrewso** [nrc-cnrc/EGSnrc#698] [Ideas]
This is the start of a WSL compatibility discussion for EGSnrc, If there are specific requests for certain functionality tests, please reply below. 

After last week's announcement of the new, Microsoft supported method of getting Bash on windows I played around with it for a bit. The great news is that all of the command line utilities appear to work, and the only flaws related to the expected lack of an x server and kde. A known issue in the current release of WSL. 

I would continue testing if not for the fact that the rest of the Windows 10 release was quite buggy, including my loosing the ability to file bug reports. I will append this issue with further testing when the WSL moves to a more stable testing branch of Windows 10. 

Again please list anything you'd like me to test or attache inp and modules. I'd like to see just how capable a prototyping system one can run on this system.

**Comment by ftessier** [comment]
Thank you @crcrewso! The advent of a native bash shell and the ability to run ELF binaries directly within the Windows Subsystem for Linux bodes very well for EGSnrc on Windows! We are very interested to hear more on your testing efforts, please keep us up to date. We are looking forward to maintaining a single set of files and configuration scripts on all platforms.

**Comment by crcrewso** [comment]
The Fast ring is a bit unstable for the rest of my work and the known nano issue with the current WSL system is driving the current fast ring community a bit bonkers (I know I know, not using vim is a sin) so I'm going to delay further testing until the summer release of the stable environment. I'll keep checking on this thread if people want to submit specific test requests.

**Comment by ftessier** [comment]
Sounds good, we are looking forward to hearing back on the stable release!

**Comment by crcrewso** [comment]
The stable release is out. I've spent about a day testing and found a few things. 

1) You can get GUI's working with some effort
2) Command line stuff works as expected, but the base Ubuntu image is missing a lot of default packages which need to be installed

Preparation steps
1) Read the windows guide about [installing ](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide)the Linux subsystem for Windows. 
2) Install the WSL, 
3) apt-get install git g++ gcc gfortran make 
3) optional if you're going to be adding the gui's add tcl tk grace qtcreator and a text editor (geany and medit were my first two) 

_note_ this is where I deviated from 'normal'. I performed the unsupported and unapproved action of [updating ](http://disq.us/p/1aondlg) Ubuntu to 16.04.02 

Then I got xserver working through the [dbus ](https://www.reddit.com/r/Windows10/comments/4rsmzp/bash_on_windows_getting_dbus_and_x_server_working/?ref=share&ref_source=link) hack which I tested with medit and xclock. 

Once that was all squared away I was able to fetch the Develop branch of EGSnrc and build it using the EGSnrc_configure_linux utility.

I've also installed it from this point using HEN_HOUSE/scripts/configure, which is why I put in that feature request today. 

I'm still testing and will write up something a bit better. I just wanted to get the basics down quickly.

**Comment by ftessier** [comment]
@crcrewso do you have any update on configuring EGSnrc under the Windows Subsytem for Linux (WSL)?

**Comment by crcrewso** [comment]
I don't really. If you treat WSL like ssl'ing into a server without X then our documentation matches the capabilities. Where it breaks down is with getting X and the guis working. All of these issues have solutions in the WSL community, but they change between versions. 

There is a new version of the WSL coming out in the Windows 10 Creators Update. I am hoping that after testing that version will be a LOT more user friendly.

**Comment by ftessier** [comment]
So the EGSnrc configuration runs on WSL? What about running simulations, does that work as expected? If only the guis are not working, then we ought to start documenting this in the Windows install Wiki. Presumably one can use the pre-compiled guis for Windows in conjunction with a WSL install?

**Comment by crcrewso** [comment]
@ftessier I'd be happy to help with that, one suggestion would be that the documentation not go live until after the Creators Update, as they are finally launching a newer version of Ubuntu than 14.04. 

The simulations do run as expected, I've only been able to validate results to a point as I was using it to design my scripts and input files but I was actually running them for result on clusters. So I was running 100,000 histories on my little intel 6Y30 (or the desktop 4 core i5) and rerunning at 100,000,000 on a cluster

What would you like to see from me to help with this documenation?

**Comment by ftessier** [comment]
Sounds good. If you want, create a new mardown wiki page called "Install EGSnrc on Windows WSL", and we can add it to the wiki. If you want to start from the existing Windows install wiki page, simply clone the wiki with `git clone https://github.com/nrc-cnrc/EGSnrc.wiki.git`. You won't be able to push but you can attach the file to a comment and we can include it (with acknowledgement).

**Comment by crcrewso** [comment]
I have a page started that I'll be working from and testing once the Creator's Update launches. I should be able to send the procedure back to you after a couple of weeks of testing. 

At present it's written from the point of view of a Linux user who has to install on a windows machine. Should I spend time expanding it to 'a window's user who has seen a tiny bit of Linux'?

**Comment by crcrewso** [comment]
I thought I'd update this thread. the reason for my delay in submitting the notes is twofold. Ubuntu has not been totally clear on how they will depreciate the x11 packages in the archive through the 18.04.  This affects using vcxsrv to display our GUI's (all work right now). Additionally WSL is just now coming out of Beta. I will shoot for getting this fully documented for the Spring Creators Update so it 'should' be fully tested for the 2019 release. 

I hope to include documentation for the Ubuntu/Debian and Fedora releases (if Fedora ever launches) Is it worth my time to look at Suse too?

**Comment by crcrewso** [comment]
I have found a complete use stopping bug in the WSL layer. The power management 'features' they added at some point seem to cause recent branches of EGS to freeze if the OS think's you're not using active WSL apps. I'm trying to track it down but at this point I've had to revert back to VM's for runs longer than a couple minutes.

**Reply by crcrewso** [reply]
This issue seems to have been solved. I would recommend only supporting WSL2 on 2004 or 20H2+ windows 10 installs.

**Reply by ftessier** [reply]
Thanks for the update!

**Comment by ftessier** [comment]
See update in #767 as well.

**Comment by crcrewso** [comment]
I've done quite a bit of work on this and am using WSL in the office with great success. 

In short:

- Install WSL2
- Treat as linux using linux documentation


I've tried every linux distro I could to this point, and they all work as expected. I'd suggest defaulting to Ubuntu unless one needs something different. In all cases I made sure to install: make, cmake, tcl, tcl-dev, tk, tk-dev, gcc, gfortran, x11-apps, qt5-default, qt5-qmake, and grace. the x11-apps and qt5-default are required to get the prerequisites for the qt guis loaded on the default Ubuntu install since it's currently configured as a headless server. This will probably change with the 22.04 version provided through the store. 

I would also recommend 
- Installing [VS Code](https://code.visualstudio.com/) and Microsoft Terminal (from the store or soon to be default within W11)
- Read [this getting started guide](https://code.visualstudio.com/docs/remote/wsl) to use Visual Studio Code to easily edit egsinp files for simulations

For Windows 11 downloading the WSL2 install from the windows store does everything including provides gui support. 
[Documentation](https://docs.microsoft.com/en-us/windows/wsl/install#ways-to-run-multiple-linux-distributions-with-wsl)
[Link](https://aka.ms/wslstore)
This requires relatively recent gpu drivers, so if it's not working try one of the drivers linked [here](https://github.com/microsoft/wslg#pre-requisites)

For Windows 10, installing as described above works perfectly, to get guis working one needs to add some form of X Server. Installing [gWSL](https://opticos.github.io/gwsl/) from the Windows store might be the easiest to support. 

In all cases I would consider the documentation already available to be of more than sufficient quality to use as a base, instead of writing our own specific guide to getting WSL started. While certainly production ready, each vendor is doing what they can to simplify this process and the Microsoft documentation is updated to keep pace. 

With the openSSH client being now built into the Microsoft Terminal I can't see the historical tools for getting EGSnrc up and running on Windows being well supported for much longer.
